### PR TITLE
Add permissions patches

### PR DESCRIPTION
**Pull request by blu25**
***
<a href="https://pr-preview.s3.amazonaws.com/WICG/fenced-frame/pull/74.html" title="Last updated on Apr 30, 2023, 12:35 PM UTC (cad5853)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/fenced-frame/74/0582b0e...cad5853.html" title="Last updated on Apr 30, 2023, 12:35 PM UTC (cad5853)">Diff</a>

### CODE REVIEW

**Comment by blu25** [comment]
> Is this ready for my review right now or is it still in draft shape?

This is ready for your review, but is not ready to be taken out of draft. There are a couple TODOs still in the patches that I need your input on how to handle.

File: spec.bs
Comment (domfarolino) [inline spec.bs:347]: In this PR there is no content attribute. You'll need to add an attribute to this section: https://github.com/WICG/fenced-frame/blob/master/spec.bs#L229

File: spec.bs
Comment (domfarolino) [inline spec.bs:130]: This does not seem used?

File: spec.bs
Comment (domfarolino) [inline spec.bs:1609]: What's up with this?

**Comment by domfarolino** [comment]
> There are a couple TODOs still in the patches that I need your input on how to handle.

Can you be more specific with what you need help on?

File: spec.bs
Comment (domfarolino) [inline spec.bs:1651]: `<{iframe}>` and `<{fencedframe}>`, per https://speced.github.io/bikeshed/#metadata-markup-shorthands

Comment (domfarolino) [inline spec.bs:1651]: "allow" should be a real link to the associated attributes

File: spec.bs
Comment (domfarolino) [inline spec.bs:1689]: Let's briefly mention that these changes are made _in addition to_ the changes made in https://wicg.github.io/fenced-frame/#creating-browsing-contexts-patch (please use https://speced.github.io/bikeshed/#section-links to link to that section).

File: spec.bs
Comment (domfarolino) [inline spec.bs:1651]: Same here, and elsewhere if applicable, regarding the element and attribute links.

File: spec.bs
Comment (domfarolino) [inline spec.bs:1628]: Is this step 4 of the new algorithm (with the above changes) or of the existing algorithm not considering the changes above.

File: spec.bs
Comment (blu25) [reply spec.bs:1609]: I think I just forgot to remove that line.

File: spec.bs
Comment (blu25) [reply spec.bs:1628]: Step 4 of the existing algorithm. I'm not sure what the convention is for referring to steps after new steps were added before it in a patch.

File: spec.bs
Comment (domfarolino) [inline spec.bs:1684]: Just a heads up this isn't actually an algorithm. If you look under this heading in the HTML Standard, it references the ["create and initialize a Document object"](https://html.spec.whatwg.org/C#initialise-the-document-object) algorithm, which is what we should be referencing.

**Comment by domfarolino** [comment]
> I'm not entirely sure how to get access to the fenced frame config.

Let me make sure I understand: these are cases where you are effectively "inside" the fenced frame and you need to access i.e., the "fenced frame properties" implementation-equivalent object in the spec, right? (Basically what we have in DocumentLoader in our implementation inside the FF renderer?) In that case, I think you can do what we do in https://wicg.github.io/fenced-frame/#ref-for-fenced-frame-config-instance%E2%91%A2, including with the XXX box below until I fix that :/

Does that help?

File: spec.bs
Comment (domfarolino) [inline spec.bs:1713]: So this is the algorithm that sees whether or not a fenced frame has all of the permissions that the "config instance" (read: fenced frame properties) demands that it have, right? In that case, I think we should not have this be called by Fetch, but instead basically be one of the many algorithms that we call before we finalize the navigation, that can lead us down the error path. See for example: https://wicg.github.io/fenced-frame/#ref-for-attempt-to-populate-the-history-entry's-document, which basically references algorithms like https://html.spec.whatwg.org/multipage/browsing-the-web.html#attempt-to-populate-the-history-entry's-document:~:text=Otherwise%2C%20if%20the%20result%20of%20checking%20a%20navigation%20response%27s%20adherence%20to%20%60X%2DFrame%2DOptions%60, whose placement I think we want to mimic with the `S-L-M` header checks and probably the permissions policy checks as well, especially since these are navigation-specific checks.

**Comment by domfarolino** [comment]
Please feel free to mark all previous discussions has "resolved" with the GitHub UI just so I'm sure which ones still need attention vs which ones are indeed addressed.

File: spec.bs
Comment (gtanzer) [inline spec.bs:644]: We don't need this field (yet). This is the point of [=fenced frame config/effective permissions=].

"required permissions to load" as a name doesn't capture the fact that it disables all other permissions anyway.

File: spec.bs
Comment (domfarolino) [inline spec.bs:1816]: All of these references to container policy should probably be updated to point to either the already-existing "container policy" or the new "fenced container policy" when appropriate. This all assumes that we should introduce the concept of a "fenced container policy" that's independent of "container policy", instead of just modifying the existing "container policy definition accordingly. I'll let you decide which is best to do.

File: spec.bs
Comment (domfarolino) [inline spec.bs:1804]: I think this is broken as-is right now. Your later modifications to https://html.spec.whatwg.org/multipage/document-sequences.html#creating-a-new-browsing-context pass in nothing as the `|required policies|` variable, but it seems that when `|container|` is a fenced navigable container, we expect `|required features|`. We should fix this and in the process decide whether:
 - The initial `about:blank` Document in a fenced frame should have _all_ policies manually disabled; or...
 - Or inherit them from "beyond the fence" as iframes would normally do

The first option seems like the safest, and is kinda not web-observable except for with extensions I guess. But I'd love to know what our implementation does.

File: spec.bs
Comment (domfarolino) [inline spec.bs:1885]: Is the intention of this algorithm really to not consider any of the permissions policies in the response? My understanding is that we wanted to make sure that the required permissions were not blocked by either:
 1. A parent document's permissions that were inherited in the disabled state
 2. Response headers in the actual response

Do we not care about (2) above? Is the only piece of the response that we need, the origin?

File: spec.bs
Comment (domfarolino) [inline spec.bs:1901]: My understanding is that `inherited policy` will be completely full of all supported features, so it should always exist, per https://w3c.github.io/webappsec-permissions-policy/#inherited-policy-for-a-feature.

Do you mean to check if `|inherited policy|[|permission|] is "Disabled"` instead?

File: spec.bs
Comment (blu25) [reply spec.bs:1804]: I agree with doing option 1. That will match the behavior in our implementation once the required permissions CL is landed. If an empty permissions list is passed into PermissionsPolicy::CreateForFencedFrame, it will disable every feature.

File: spec.bs
Comment (blu25) [reply spec.bs:1884]: Re: response permissions policies, [we have a WPT](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/web_tests/external/wpt/fenced-frame/default-enabled-features-attribute-allow.https.html;l=25-35;drc=3ccc9b67e688ec5f98f6664385f0587fb724d307) to check that we can further restrict permissions after a fenced frame loads

Re: the wildcard, [our implementation](https://source.chromium.org/chromium/chromium/src/+/main:content/browser/renderer_host/navigation_request.cc;l=8367-8384;drc=dfb15f3bc92332c4f433ad9721ba1a16ce2595cc) matches what's in the spec.
`the policy must be either explicitly enabled for 'origin', or the policy must by default be enabled for all origins`

File: spec.bs
Comment (blu25) [reply spec.bs:1832]: My vote is to stick with "container policy". The container policies themselves behave the exact same between fenced frames and iframes. The difference is how external algorithms handle them. So I think we'd just be making a duplicate of "container policy" with "fenced container policy".

File: spec.bs
Comment (blu25) [reply spec.bs:1778]: I'm going to stick with the existing container policy definition. I don't think we need a new "fenced container policy" here.

File: spec.bs
Comment (domfarolino) [inline spec.bs:1888]: We shouldn't get the response's origin, but rather the navigation params's origin (https://html.spec.whatwg.org/C#navigation-params-origin), as this will be the actual origin used for the document.

File: spec.bs
Comment (gtanzer) [inline spec.bs:1799]: extra 's on <{fencedframe}>'s here

File: spec.bs
Comment (gtanzer) [inline spec.bs:1782]: This is true right now, but maybe not true with future changes (or default mode, depending on how we spec it).

File: spec.bs
Comment (gtanzer) [inline spec.bs:1831]: `{Document}}` -> `{{Document}}`

File: spec.bs
Comment (domfarolino) [reply spec.bs:1782]: Right, thanks. For now I'm just worried about spec'ing opaque stuff basically.

File: spec.bs
Comment (domfarolino) [inline spec.bs:1957]: Do we need to also modify this algorithm to "fence" the container policy look-up? Right now https://w3c.github.io/webappsec-permissions-policy/#define-inherited-policy-in-container looks at the container policy (which may contain 'none', `self', 'src`, '*', or an origin) and runs the "matches" algorithm there. I think we'll need to ensure that 'self', 'src', and "an origin" are "fenced" and fail right? Or is there a reason we don't have to do this?

**Review by domfarolino — APPROVED** [review]
I added some more fixes and editorial content. I think this is all ready to go % one single comment!

File: spec.bs
Comment (domfarolino) [reply spec.bs:1957]: Filed https://github.com/WICG/fenced-frame/issues/82 so we can just land this PR now.

**Pull request merged**
